### PR TITLE
Bug-fixing in dual averaging for HMC

### PR DESCRIPTION
**Pull request by yebai**
The step-size adaption problem is probably resolved in this PR. So far I have been using the `gdemo` example, and the `m12_2` model from `StatisticalRethinking` for testing. Please take a look and let me know if you have better models for testing.

The effective sample size for `m12_2` 

- master branch: https://github.com/TuringLang/Turing.jl/pull/641#issuecomment-456224538
- this PR: https://github.com/TuringLang/Turing.jl/pull/641#issuecomment-471019628

I'm working on a more robust version of `_find_good_eps`. When this is done, the following line will be changed back to ` ϵ = direction == 1 ? 2 * ϵ : 0.5 * ϵ `

https://github.com/TuringLang/Turing.jl/blob/746443249e8ebfe3ede93da3a515ce78403205af/src/inference/support/hmc_core.jl#L269

Related: #628

### CODE REVIEW

**Comment by yebai** [comment]
Models being used

`gdemo`

```julia
using Revise, Turing

@model gdemo(x, y) = begin
  s ~ InverseGamma(2,3)
  m ~ Normal(0,sqrt(s))
  x ~ Normal(m, sqrt(s))
  y ~ Normal(m, sqrt(s))
end

c5 = sample(gdemo(1.5, 2), HMCDA(2000, 0.15, 0.65))
```


`m12_2`

```julia
using StatisticalRethinking
using Turing

Turing.setadbackend(:reverse_diff)

d = CSV.read(joinpath(dirname(Base.pathof(StatisticalRethinking)), "..", "data",
    "reedfrogs.csv"), delim=';')
size(d) # Should be 48x5

# Set number of tanks
d[:tank] = 1:size(d,1)

@model m12_2(density, tank, surv) = begin
    # Separate priors on α and σ for each tank
    σ ~ Truncated(Cauchy(0, 1), 0, Inf)
    α ~ Normal(0, 1) #Check stancode this might need to be in the for loop below?

    # Number of unique tanks in the data set
    N_tank = length(tank)

    # Set an TArray for the priors/param
    # a_tank = TArray{Any}(undef, N_tank)
    a_tank = Vector{Real}(undef, N_tank)

    # For each tank [1,..,48] set a prior
    for i = 1:N_tank
        a_tank[i] ~ Normal(α, σ)
    end

    for i = 1:N_tank
        logitp = a_tank[tank[i]]
        surv[i] ~ BinomialLogit(density[i], logitp)
    end
end

posterior = sample(m12_2(d[:density], d[:tank], d[:surv]),
                   Turing.NUTS(4000, 500, 0.95),
                   # Turing.HMCDA(4000, 500, 0.95, 0.3)
                  )
describe(posterior)
```

**Comment by yebai** [comment]
Update:

Running the LDA example will occasionally throw `log(0)` error, we need to catch this type of exception / error in Turing, and reject the current HMC proposal when it happens. This is likely to require some re-work of the error handling mechanism proposed in https://github.com/TuringLang/AdvancedHMC.jl/issues/16

```julia
julia> chain = sample(mf, HMCDA(2000, 0.65, 1.0))
[ Info: [Turing] looking for good initial eps...
[ Info: [Turing] ϵ = 0.2, accept ratio a = 6.58293101371248e-25
[ Info: [Turing] ϵ = 0.15000000000000002, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.17500000000000002, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.1875, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.19375, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.19687500000000002, accept ratio a = 1.6572427130567486e-9
[ Info: [Turing] ϵ = 0.1953125, accept ratio a = 0.04197809303831285
[ Info: [Turing] ϵ = 0.19453125, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.194921875, accept ratio a = 1.0
[ Info: [Turing] ϵ = 0.1951171875, accept ratio a = 0.3421968454472759
[ Info: [Turing] found initial ϵ: 0.1951171875
ERROR: DomainError with -2.220446049250313e-16:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:  1.0
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:31
 [2] log(::Float64) at ./special/log.jl:285
 [3] log(::ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Float64,10}) at /Users/hg344/.julia/packages/ForwardDiff/N0wMF/src/dual.jl:203
 [4] logpdf_with_trans(::Dirichlet{Float64}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Real,10},1}, ::Bool) at /Users/hg344/.julia/packages/Bijectors/vCpPQ/src/Bijectors.jl:292
 [5] assume(::Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}, ::Dirichlet{Float64}, ::Turing.Core.VarReplay.VarName, ::Turing.Core.VarReplay.VarInfo) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:277
 [6] macro expansion at /Users/hg344/.julia/dev/Turing/src/core/compiler.jl:102 [inlined]
 [7] macro expansion at ./REPL[27]:9 [inlined]
 [8] (::getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}})(::Turing.Core.VarReplay.VarInfo, ::Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}, ::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}) at /Users/hg344/.julia/dev/Turing/src/core/compiler.jl:389
 [9] #call#3 at /Users/hg344/.julia/dev/Turing/src/Turing.jl:62 [inlined]
 [10] Model at /Users/hg344/.julia/dev/Turing/src/Turing.jl:62 [inlined]
 [11] runmodel!(::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}, ::Turing.Core.VarReplay.VarInfo, ::Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}) at /Users/hg344/.julia/dev/Turing/src/core/VarReplay.jl:110
 [12] f at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:109 [inlined]
 [13] chunk_mode_gradient!(::Array{Real,1}, ::getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}, ::Array{Real,1}, ::ForwardDiff.GradientConfig{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Real,10,Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Real,10},1}}) at /Users/hg344/.julia/packages/ForwardDiff/N0wMF/src/gradient.jl:140
 [14] gradient! at /Users/hg344/.julia/packages/ForwardDiff/N0wMF/src/gradient.jl:37 [inlined]
 [15] gradient!(::Array{Real,1}, ::getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}, ::Array{Real,1}, ::ForwardDiff.GradientConfig{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Real,10,Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing.Core, Symbol("#f#26")){Turing.Core.VarReplay.VarInfo,Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}},Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},Real},Real,10},1}}) at /Users/hg344/.julia/packages/ForwardDiff/N0wMF/src/gradient.jl:33
 [16] gradient_logp_forward(::Array{Real,1}, ::Turing.Core.VarReplay.VarInfo, ::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}, ::Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}) at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:116
 [17] gradient_logp at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:80 [inlined]
 [18] #2 at /Users/hg344/.julia/dev/Turing/src/inference/support/hmc_core.jl:12 [inlined]
 [19] #_leapfrog#26(::getfield(Turing.Inference, Symbol("##6#7")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}, ::getfield(Turing.Inference, Symbol("##8#9")){Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}, ::Function, ::Array{Real,1}, ::Array{Float64,1}, ::Int64, ::Float64, ::getfield(Turing.Inference, Symbol("##2#3")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}},Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}}) at /Users/hg344/.julia/dev/Turing/src/inference/support/hmc_core.jl:158
 [20] (::getfield(Turing.Inference, Symbol("#kw##_leapfrog")))(::NamedTuple{(:rev_func, :log_func),Tuple{getfield(Turing.Inference, Symbol("##6#7")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},getfield(Turing.Inference, Symbol("##8#9")){Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}}}, ::typeof(Turing.Inference._leapfrog), ::Array{Real,1}, ::Array{Float64,1}, ::Int64, ::Float64, ::Function) at ./none:0
 [21] #_hmc_step#27(::Function, ::Function, ::Function, ::Array{Real,1}, ::Float64, ::getfield(Turing.Inference, Symbol("##4#5")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}},Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}}, ::Function, ::getfield(Turing.Inference, Symbol("##20#21")){Array{Float64,1}}, ::Int64, ::Float64, ::getfield(Turing.Inference, Symbol("##18#19")){Int64,Array{Float64,1}}) at /Users/hg344/.julia/dev/Turing/src/inference/support/hmc_core.jl:195
 [22] (::getfield(Turing.Inference, Symbol("#kw##_hmc_step")))(::NamedTuple{(:rev_func, :log_func),Tuple{getfield(Turing.Inference, Symbol("##6#7")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},getfield(Turing.Inference, Symbol("##8#9")){Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}}}, ::typeof(Turing.Inference._hmc_step), ::Array{Real,1}, ::Float64, ::Function, ::Function, ::Function, ::Int64, ::Float64, ::getfield(Turing.Inference, Symbol("##18#19")){Int64,Array{Float64,1}}) at ./none:0
 [23] #_hmc_step#28 at /Users/hg344/.julia/dev/Turing/src/inference/support/hmc_core.jl:224 [inlined]
 [24] #_hmc_step at ./none:0 [inlined]
 [25] #hmc_step#30 at /Users/hg344/.julia/dev/Turing/src/inference/hmcda.jl:72 [inlined]
 [26] (::getfield(Turing.Inference, Symbol("#kw##hmc_step")))(::NamedTuple{(:rev_func, :log_func),Tuple{getfield(Turing.Inference, Symbol("##6#7")){Turing.Core.VarReplay.VarInfo,Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}},getfield(Turing.Inference, Symbol("##8#9")){Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}}}}, ::typeof(Turing.Inference.hmc_step), ::Array{Real,1}, ::Float64, ::Function, ::Function, ::Function, ::Float64, ::HMCDA{Turing.Core.ForwardDiffAD{40},Any}, ::getfield(Turing.Inference, Symbol("##18#19")){Int64,Array{Float64,1}}) at ./none:0
 [27] step(::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}, ::Turing.Sampler{HMCDA{Turing.Core.ForwardDiffAD{40},Any}}, ::Turing.Core.VarReplay.VarInfo, ::Val{false}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:236
 [28] macro expansion at ./util.jl:213 [inlined]
 [29] #sample#37(::Bool, ::Nothing, ::Int64, ::Nothing, ::Function, ::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}, ::HMCDA{Turing.Core.ForwardDiffAD{40},Any}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:153
 [30] sample(::Turing.Model{Tuple{:theta,:phi},Tuple{:w},getfield(Main, Symbol("###inner_function#413#22")){Int64,Int64,Int64,Int64,Array{Int64,1},Array{Float64,1},Array{Float64,1}},NamedTuple{(:w,),Tuple{Array{Int64,1}}},NamedTuple{(:w,),Tuple{Symbol}}}, ::HMCDA{Turing.Core.ForwardDiffAD{40},Any}) at /Users/hg344/.julia/dev/Turing/src/inference/hmc.jl:108
 [31] top-level scope at none:0
```

**Comment by yebai** [comment]
Ready for a review now. The remaining issue is not step-size related and better be addressed in a separate PR.

File: src/core/ad.jl [hunk 1]
```diff
@@ -166,7 +166,7 @@ end
 
 function verifygrad(grad::AbstractVector{<:Real})
     if any(isnan, grad) || any(isinf, grad)
-        @warn("Numerical error has been found in gradients.")
+        @warn("Numerical error in gradients. Rejectting current proposal...")
```
Comment (xukai92) [inline src/core/ad.jl]: Rejectting -> Rejecting

File: src/inference/adapt/stepsize.jl [hunk 2]
```diff
@@ -87,15 +87,16 @@ function adapt_stepsize!(da::DualAveraging, stats::Real)
     η_x = m^(-κ)
     x_bar = (1.0 - η_x) * x_bar + η_x * x
 
-    ϵ = exp(x)
+    ϵ = exp(x_bar)
```
Comment (xukai92) [inline src/inference/adapt/stepsize.jl]: Can you explain this change a bit more? I re-read the NUTS paper (Algorithm 6 of https://arxiv.org/pdf/1111.4246.pdf) and the Stan code (https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/stepsize_adaptation.hpp#L83) - both of them seem to do what was done in Turing before (i.e. `ϵ = exp(x)`). What's exactly going on here?

File: src/inference/support/hmc_core.jl [hunk 3]
```diff
@@ -232,53 +234,60 @@ function find_good_eps(model, spl::Sampler{T}, vi::VarInfo) where T
...
+    @info "[Turing] found initial ϵ: $ϵ"
     return ϵ
 end
 
-function _find_good_eps(θ, lj_func, grad_func, H_func, momentum_sampler; max_num_iters=12)
+function _find_good_eps(θ, lj_func, grad_func, H_func, momentum_sampler; max_num_iters=100)
     @info "[Turing] looking for good initial eps..."
-    ϵ = 0.1
+    ϵ_prime = ϵ = 0.1
+    a_min, a_cross, a_max = 0.25, 0.5, 0.75
```
Comment (xukai92) [inline src/inference/support/hmc_core.jl]: What do these variables mean? Maybe put a comment.

File: src/inference/support/hmc_core.jl [hunk 4]
```diff
@@ -232,53 +234,60 @@ function find_good_eps(model, spl::Sampler{T}, vi::VarInfo) where T
...
         end
-
-        iter_num += 1
     end
 
-    while h == Inf  # revert if the last change is too big
-        ϵ = ϵ / 2               # safe is more important than large
-        θ_prime, p_prime, τ = _leapfrog(θ, p, 1, ϵ, grad_func)
+    ϵ, ϵ_prime = ϵ < ϵ_prime ? (ϵ, ϵ_prime) : (ϵ_prime, ϵ) # Ensure ϵ < ϵ_prime
+    for _ = 1:max_num_iters
```
Comment (xukai92) [inline src/inference/support/hmc_core.jl]: Is this doing a binary search to find the largest `ϵ` between `ϵ` and `ϵ_prime` which doesn't throw numerical error?

File: src/inference/adapt/stepsize.jl (hunk 2 above)
Comment (yebai) [reply src/inference/adapt/stepsize.jl]: sorry; the old version is correct -  I was confused by `ϵ_NAdapt` and `ϵ`. The actual bug was fixed by 

https://github.com/TuringLang/Turing.jl/blob/c27e88f9dc1d00089021491fe62a9d1c85072137/src/inference/adapt/stepsize.jl#L99

ESS for `m12_2` gets notificablely better than current after reverting this to `ϵ = exp(x)`.

File: src/inference/support/hmc_core.jl (hunk 4 above)
Comment (yebai) [reply src/inference/support/hmc_core.jl]: sort of - this step ensures the final accept ratio is in the interval `[a_min, a_max]`

File: src/inference/support/hmc_core.jl (hunk 3 above)
Comment (yebai) [reply src/inference/support/hmc_core.jl]: Solved by 8486578

File: src/inference/support/hmc_core.jl [hunk 5]
```diff
@@ -232,53 +234,65 @@ function find_good_eps(model, spl::Sampler{T}, vi::VarInfo) where T
...
-    delta_H = H0 - h
-    direction = delta_H > log(0.8) ? 1 : -1
-
-    iter_num = 1
-
-    # Heuristically find optimal ϵ
-    while (iter_num <= max_num_iters)
-        θ = θ_prime
+    delta_H = H0 - h # logp(θ`) - logp(θ)
+    direction = delta_H > log(a_cross) ? 1 : -1
```
Comment (trappmartin) [inline src/inference/support/hmc_core.jl]: Nice!

**Pull request merged**